2018/burton2/tac.1
.TH tac 1 "01 April 2018" "tac" "IOCCC 2018"
.SH NAME
tac \- tokenize and count C-like program source code
.SH SYNOPSIS
\fBtac\fP [\-t] [\-c] [\-k] [\-r] [\-s] [\-i] [file]
.SH DESCRIPTION
\fBtac\fP reads C or C-like source code from \fBstdin\fP or file, breaking it into tokens (lexemes),
counting these according to the rules of the IOCCC, and outputting the tokens,
counts, or source code as directed by the command line options.
The default output is the source code with comments removed on \fBstdout\fP,
and the counts on \fBstderr\fP.
The count format is a set of integer values in the following order:
.RS
.TP
.B lines
number of newlines
.TQ
.B words
number of words
.TQ
.B chars
number of characters
.TQ
.B ioccc
number of IOCCC countable tokens; the IOCCC rule 2 value
.TQ
.B saved
number of characters not counted due to rules and keywords
.TQ
.B kwcnt
number of keywords
.TQ
.B kwlen
number of characters saved by counting keyword length as 1
.RE
.PP
For the following code:
.sp
.RS 4
.nf
#include <stdio.h>
int main() {
    /* example for loop */
    for (int ac=1; ac<10; ++ac)
	    printf("%d\n", ac);
    return 0;
}
.fi
.RE
\fBtac\fP prints the code to \fBstdout\fP, and to \fBstderr\fP: '6 15 96 71 25 5 0'
which means 6 lines, 15 words, 96 chars, 71 IOCCC counts, 25 chars saved,
5 \fBkwcnt\fP, and no characters saved from \fBkwlen\fP. (By default, comments are not
counted.)
.SH OPTIONS
The following options are supported:
.TP
.B \-t
Output just the token stream.
.TP
.B \-c
Correct the accounting bugs in \fBiocccsize(1)\fP, output the token id and the token.
.TP
.B \-k
Keep the comments.
.TP
.B \-r
Reserved words count as size 1.
.TP
.B \-s
Suppress the source code, summarize counts only.
.TP
.B \-i
Report only the official IOCCC rule 2 size: number of octets (implies \fB\-r\fP and \fB\-s\fP).
.PP
Counts are impacted by the \fB\-k\fP, \fB\-r\fP and \fB\-c\fP options.
By default, the counts do not include comments, words are counted as in \fBwc(1)\fP, and reserved words count in full.
Using the example code above, the options change the counts as follows:
.PP
.TP
\fB\-r\fP
6 15 96 53 25 5 18
.TQ
\fB\-k\fP
7 20 120 89 31 5 0
.TQ
\fB\-c\fP
7 36 98 71 27 5 0
.PP
The \fB\-c\fP option counts tokens instead of words, and is therefore more accurate for source code.
Note the aberration for line count with \fB\-r\fP; it is as\-if the comment never existed.
.PP
The \fB\-c\fP and \fB\-t\fP options change the output, instead writing a stream of tokens to \fBstdout\fP, one token per line.
The \fB\-t\fP option suppresses both the counts and the token ids, writing only the tokens.
.SH EXAMPLES
.PP
Summarize the source code counts:
.sp
.RS
.ft B
tac -rs tac.c
.ft R
.RE
.PP
Count the 10 most frequently used unique tokens in the source code:
.sp
.RS
.ft B
tac -t tac.c 2>/dev/null | sort | uniq -c | sort -k1nr | sed 10q
.ft R
.RE
.PP
Count the unique constants in the source:
.sp
.RS
.ft B
tac -c tac.c 2>/dev/null | grep ^300 | sort | uniq -c | sort -k1nr | awk '{print $1,$3}'
.ft R
.RE
.SH BUGS
\fBtac\fP does not properly count preprocessor tokens with embedded space or comments as reserved words:
.sp
.RS
.ft B
echo '#/*comment*/include/*comment*/<stdio.h>' | tac -i
.br
echo '# include<stdio.h>' | tac -i
.br
echo '#include<stdio.h>' | tac -i
.ft R
.RE
.sp
report no reserved words on the first two inputs, and thus 17 as the IOCCC size, although 10 is the correct answer (as in the final example).
.PP
The original meaning of the \fB\-c\fP option was "become compatible with \fBiocccsize(1)\fP".
Its meaning was flipped so the default behavior is now compatible.
.SH SEE ALSO
\fBiocccsize(1)\fP, \fBwc(1)\fP
.br
\<https://ioccc.org/2015/rules.txt\>


news.md
# For more news, try mastodon

### Please follow the [IOCCC on Mastodon](https://fosstodon.org/@ioccc)

We use Mastodon to make minor announcements, in addition to announcing
major events such as the opening of a new IOCCC, or who won.

See the
FAQ on "[Mastodon](faq.html#try_mastodon)"
for more information on Mastodon.

<div id="news">
# News
</div>


## 2024-12-31

We have just released [version 2.3 of the mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry/releases/tag/v2.3).
This repo 2.3 release includes important improvements (over repo release 2.1) to the `mkiocccentry(1)` tool
as well as a critical fix for installing on BSD based systems.

Please download [version 2.3 of the mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry/releases/tag/v2.3),
and compile the [mkiocccentry tool set](faq.html#mkiocccentry).  You may also wish to install this code.
See also the [mkiocccentry FAQ](https://github.com/ioccc-src/mkiocccentry/blob/master/FAQ.md) for more information.


## 2024-12-30

The IOCCC [Rule 13](next/rules.html#rule13) no longer discourages the
use of UTF-8 in C source.  Since the days of `-std=gnu17` compilers,
UTF-8 characters do not impose a compiler problem.  The related guideline
states that "Rule 13 no longer discourages the use of UTF-8 +characters in C code".

The IOCCC [Rule 8](next/rules.html#rule8) wording has been fixed.  It no longer
states that email will be sent.  Instead it talks about potential slot comment
modifications as means to tell the submitter that their submission was received.

All of the import IOCCC dates within the [IOCCC rules](next/rules.html)
have been moved into the "[Important IOCCC dates](next/rules.html#dates)"
section.


## 2024-12-28

### Good News Everyone! :-)

After a 4 year effort by a [number of people](thanks-for-help.html),
with over 6168+ commits, the **Great Fork Merge** has been completed
and the [Official IOCCC web site](index.html) has been updated!

A significant number of improvements has been made to the
[IOCCC winning entries](years.html).
A number of fixes and improvements involve the ability of reasonable modern
Unix/Linux systems to be able to compile and even run them.
See [FAQ Section 4](faq.html#compiling) for
information on how to compile.
See the [FAQ Section 8](faq.html#changes)
for details on changes made to the entries.

We hope you enjoy the new and improved
[Official IOCCC web site](index.html).
This web site is backed by the
[IOCCC winner repo](https://github.com/ioccc-src/winner).
See [FAQ Section 9](faq.html#help) for
how you help [improve / fix the we site](faq.html#fix_website),
or fix [known bugs in IOCCC winning entries](bugs.html).


### IOCCC28 is starting!

The IOCCC28, the [40th anniversary of the IOCCC](faq.html#ioccc_start) is now
scheduled to run from **2024-12-29 23:58:13.213455 UTC** to **2025-04-01 23:29:31.374143 UTC**!!!.

The [contest status](status.html) as been changed to "**pending**"
and [status.json](status.json) has been updated to reflect this new status.

Proposed and tentative [IOCCC rules](next/rules.html) and [IOCCC guidelines](next/guidelines.html)
have been released.

See [FAQ Section 0](faq.html#enter_questions) for more information on how to enter.


## 2024-06-08

We continue to make good progress on the
[temp-test-ioccc website](https://ioccc-src.github.io/temp-test-ioccc/)
via the [temp-test-ioccc](https://github.com/ioccc-src/temp-test-ioccc) repo.

One of the [largest html
files](thanks-for-help.html) has
been checked for broken links which leaves two longer files to go. These two
files are the [bugs.html](bugs.html)
file and the "[IOCCC FAQ](faq.html)".

However, where it was thought that those two files would be looked at before the
other html files, it has been decided that they will be done after the other
files as it is very possible that some things in these two files will have to be
changed or added as the other files are reviewed for broken links.

As of today, 08 June 2024, the top level html files are being looked at, with at
least three done already ([judges.html](judges.html), [news.html](news.html) and
[status.html](status.html)). This should not take more than a few days if things
go as planned, as these files are small and should be relatively quick and easy
to check for broken links. Of course, some of these, like the
[news.html](news.html) and possibly [status.html](status.html) will likely have to be
checked again but as for [news.html](news.html) the dates up to and including _08
June 2024_ have been checked for invalid links so that only new news updates
will have to be checked.

Once those are done the html files of all the entries will be looked at. As
those are being reviewed it the [bugs.html](bugs.html) might very well be
updated and the "[IOCCC FAQ](faq.html)" might be as well.  This is why, as noted above,
we have decided to review these files last, in order to save time and effort in
not having to look at them more than once on a whole, other than possible
entries that have to be added at the end, like the example below.

Once the _submit server_ server is ready another FAQ will be added and that
entry can be looked at alone.


## 2024-05-30

A number of issues from minor fixes have been applied to the
[temp-test-ioccc website](https://ioccc-src.github.io/temp-test-ioccc/)
via the [temp-test-ioccc](https://github.com/ioccc-src/temp-test-ioccc) repo.

437 commits have been applied since 2024-04-20, bringing the total
number to 5260 commits to date.  The
[temp-test-ioccc](https://github.com/ioccc-src/temp-test-ioccc)
repo is now 5159 commits ahead of the [IOCCC
winner](https://github.com/ioccc-src/winner) repo.

Multiple **((top priority))** issues have been resolved or completed.
Only one **((top priority))** issue, [issue
2006](https://github.com/ioccc-src/temp-test-ioccc/issues/2006)
remains before we begin the final steps to perform the **Great Fork
Merge** when the [Official IOCCC web

2018/burton2/tac.c
/*
 * tac - tokenize and count C-like languages
 *
 * tac is offered as a replacement for iocccsize.
 * Version 2018.04.18
 * Public Domain 2015 by Dave Burton.
 * Please keep this attribution with the source code.
 *
 * cc -O3 -DW=\"keywords\" -Wall -pedantic -o tac tac.c
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

/*
 * The keywords recognized by tac should be defined in an external file,
 * filtered by the following script into a new file, and #included
 * by defining the preprocessor variable W as this new file.
 *
 * The input to the script is a list of keywords, one word per line,
 * in any order.  The output is a single string suitable for tac.
 * --------------------------------------------------

#!/bin/sh
awk '/^[^$]/{l=length;m=m>l?m:l;a[l]=a[l]$1}
     END{printf"\"";for(i=1;i<=m;++i)printf" "a[i];print"\""}' $*

 * --------------------------------------------------
 * The pre-defined string below is the IOCCC reserved word list for 2016.
 * It was created by the above script, then inserted here in case the
 * source keywords file is not available.
 *
 * The format of the keyword string is precise; addition or removal of
 * a single character will cause improper operation.
 * 
 * DO NOT EDIT the string kw; instead:
 *
 *	mkkeywords.sh < my.list > my.keywords
 *	cc -DW=\"my.keywords\" -o tac tac.c
 */

char *keywords = ""
#ifndef W

"  ifdoor intfor#ifandnotxor voidcharelsegotocaselongenumautobooltr\
ue whilebreakconstfloatshortunion#else#elif#line_Boolfalsecomplbito\
ror_eq returnsizeofstructdoubleswitchstaticexternsigned#ifdef#endif\
inline#errorbitandand_eqnot_eqxor_eq typedefdefault#ifndef#pragma_A\
tomic_Pragmaalignasalignof #includeunsignedcontinuevolatileregister\
restrict#warning_Complex_Generic_Alignas_Alignofnoreturn _Noreturn \
_Imaginary  thread_local _Thread_localstatic_assert _Static_assert"

#else
#  include W
#  undef W
#endif
;

int cline, cword, cbyte, cioccc, ckey, cextra;			// counts
int tsaved, tkey;						// totals
int fkeep, fsuppress, freserved, fioccc, fcompat, ftoken;	// flags

unsigned char *buf, *cur, *end, *white, *start;

enum {
	  CPP   = 256
	, CMT   = 257
	, KEYW  = 258
	, DIVEQ = 259
	, MULEQ = 260
	, MODEQ = 261
	, ADDEQ = 262
	, SUBEQ = 263
	, XOREQ = 264
	, OREQ  = 265
	, ANDEQ = 266
	, RSHEQ = 267
	, LSHEQ = 268
	, NOTEQ = 269
	, EQ    = 270
	, GE    = 271
	, LE    = 272
	, OR2   = 273
	, AND2  = 274
	, PP    = 275
	, MM    = 276
	, PTR   = 277
	, LSH   = 278
	, RSH   = 279
	, INT   = 300
	, FLOAT = 301
	, ID    = 302
	, DOT3  = 303
};

enum {
	  O=0x01	// octal
	, H=0x02	// hex
	, D=0x04	// decimal
	, I=0x08	// identifier
	, W=0x10	// whitespace
	, B=0x20	// brace, semi
	, OHD = O|H|D
	, HD  = H|D
	, HI  = H|I
};

char tab[256] = {
	 W, 00, 00, 00, 00, 00, 00, 00,   00,  W,  W,  W,  W,  W, 00, 00,
	00, 00, 00, 00, 00, 00, 00, 00,   00, 00, 00, 00, 00, 00, 00, 00,
	 W, 00, 00,  I, 00, 00, 00, 00,   00, 00, 00, 00, 00, 00, 00, 00,
	OHD,OHD,OHD,OHD,OHD,OHD,OHD,OHD,  HD, HD, 00,  B, 00, 00, 00, 00,

	00, HI, HI, HI, HI, HI, HI,  I,    I,  I,  I,  I,  I,  I,  I,  I,
	 I,  I,  I,  I,  I,  I,  I,  I,    I,  I,  I, 00, 00, 00, 00,  I,
	00, HI, HI, HI, HI, HI, HI,  I,    I,  I,  I,  I,  I,  I,  I,  I,
	 I,  I,  I,  I,  I,  I,  I,  I,    I,  I,  I,  B, 00,  B, 00, 00,
};

char tri[256] = {
	['='] = '#',
	['('] = '[',
	['/'] = '\\',
	[')'] = ']',
	['\'']= '^',
	['<'] = '{',
	['!'] = '|',
	['>'] = '}',
	['-'] = '~',
};

#define is(x, c) (tab[c]&(x))

/*
 * discard backslash-newline as-if not present, even if trigraph backslash
 * iocccsize counts tri as xbcount+=2,
 * and bs-nl as xb+=2, xw++, xl++ to match wc(1)
 */

void readsource() {
	int c, d, o=2, len=4096;
	buf = malloc(len);
	while ((c = getchar()) != EOF) {
		if (c == '\r') {
			if (fcompat) ++cextra;
			continue;
		}
		if (buf[o-1] == '?' && buf[o-2] == '?' && tri[c]) {
			c = tri[c];
			o -= 2;
			if (fcompat) cextra += 2;
		}

		/*
		 * digraph support:
		 * translating digraphs after trigraphs is,
		 * for the purposes of this code, essentially during tokenization.
		 */
		d = c;
		switch (buf[o-1]) {
		case '<': d = (c == ':' ? '[' : c == '%' ? '{' : c); break;
		case '%': d = (c == '>' ? '}' : c == ':' ? '#' : c); break;
		case ':': d = (c == '>' ? ']' : c); break;
		}
		if (c != d) {
			c = d;
			o--;
			if (fcompat) ++cextra;
		}

		if (c == '\n' && buf[o-1] == '\\') {
			--o;
			cline++;
			if (fcompat) {
				++cword;
				cextra += 2;
			}
		} else
			buf[o++] = c;
		if (o == len)
			buf = realloc(buf, len+=4096);
	}
	buf[o] = 0;
	end = buf + o;
	cur = buf += 2;
}

/*
 * keyword recognition: kw is sorted by length, space separated between lengths
 *	e.g. "  ifdo forint#if charvoidelsegoto..."
 * build a table of offsets for each length
 *	e.g. [0]0, [1]1, [2]2, [3]7, [4]17, ...
 * this allows scans for only words of same length instead of entire list
 * trade-off is a small, but constant start-up cost to get a configurable
 * word list (backported from obfuscated code: it was too nice to give up)
 */

int nkwoff, kwoff[32];	   // allows keywords up to 32 characters in length

void kwinit() {
	char *k = keywords;
	for (nkwoff = 0; *k; ++k)
		while (*k == ' ')
			kwoff[++nkwoff] = ++k - keywords;
	kwoff[++nkwoff] = ++k - keywords;
}

int iskw(unsigned char *s, int len) {
	int found = 0;
	if (len < nkwoff) {
		char *off = keywords + kwoff[len];
		char *end = keywords + kwoff[len + 1] - 1;
		for (; !found && off < end; off += len)
			found = *s == *off && !strncmp((char*)s, off, len);
		if (found) {
			++ckey;
			tkey += len - 1;
		}
	}
	return found;
}

int want(char m) {
	while (cur < end) {
		if (*cur == '\\') {
			/*
			 * need to only ignore: ntbrfv\'"a?, octal, x, hex
			 * 0111 = two chars, \091 = three, \xabc = one char(!)
			 * for our purposes, only need: ' "  and \ newline,
			 * but readsource eats \ newline, so only quotes here
			 */
			if (cur[1] == '\\' || cur[1] == '\'' || cur[1] == '"') {
				cur += 2;
				continue;
			}
		}
		if (*cur == '\n') ++cline;
		if (*cur++ == m) return m;
	}
	return 0;
}

int follow(char m, int y, int n) {
	if (*cur == m) return ++cur,y;
	return n;
}

int follow2(char m, int y, int n, int y2) {
	if (*cur == m) return ++cur,y;
	if (*cur == n) return ++cur,y2;
	return n;
}

int skipwhite() {
	int c;
	for (c=*cur++; cur<=end && is(W, c); c=*cur++) {
		++cbyte;
		if (c == '\n') ++cline;
	}
	return c;
}

/*
 * CPP tokens may extend beyond a single line of text and include comments
 * To do this correctly, need to scan and remove comments during readsource
 * and present comments as whitespace.  Note that "# include" is a valid CPP
 * token, and does not match "#include".
 *
 * Because the global pointers cur, start and white are well-known, and
 * because comments are not consumed during readsource, it is more difficult
 * to parse CPP stuff.  Also note that for backwards compat, we need to also
 * count words correctly, with and without -k.
 */

int lex() {
	int c, t;

	white = cur;
	c = skipwhite();
	start = cur-1;

	if (cur >= end)
		return c;

	switch (c) {
	case '\'': want(c); return c;
	case '"':  want(c); return c;
	case '/':
		t = follow2('=', DIVEQ, c, CMT);
		if (t == CMT)
			want('\n');
		else if (t == c && (t = follow('*', CMT, c)) == CMT)
			do want('*'); while (cur<end && follow('/', 0, 1));
		return t;
	case '*':  return follow ('=', MULEQ, c);
	case '%':  return follow ('=', MODEQ, c);
	case '=':  return follow ('=', EQ,    c);
	case '!':  return follow ('=', NOTEQ, c);
	case '^':  return follow ('=', XOREQ, c);
	case '|':  return follow2('=', OREQ,  c, OR2);
	case '&':  return follow2('=', ANDEQ, c, AND2);
	case '+':  return follow2('=', ADDEQ, c, PP);
	case '-':  t =    follow2('=', SUBEQ, c, MM);
	   return t==c?   follow ('>', PTR,   t) : t;
	case '>':  t =    follow2('=', GE,    c, RSH);
	   return t==RSH? follow ('=', RSHEQ, t) : t;
	case '<':  t =    follow2('=', LE,    c, LSH);
	   return t==LSH? follow ('=', LSHEQ, t):t;

	case '#':  if (ftoken || !fcompat) return want('\n'), CPP;
	case '.':  if (cur[0] == c && cur[1] == c) return cur+=2, DOT3;
	}

	if (is(D, c)) {
	    if (c == '0' && (*cur|32) == 'x')
		c = *++cur;
	    while (is(OHD, c))
		c = *cur++;
	    if (c == '.' || (c|32) == 'e') {
		(void)strtod((char*)start, (char**)&cur);  // find end of float
		return FLOAT;
	    }
	    return --cur, INT;
	}

	if (is(I, c)) {
	    while (is(I|D, c))
	    	c = *cur++;
	    --cur;
	    return iskw(start, cur-start) ? KEYW : ID;
	}

	return c ? c : ' ';	// allows nulls in source code
}

// iocccsize compat: adjust whitespace counts
void compat_words(int tlen, int inword) {
	int c;
	for (c=0; c<tlen; ++c) {
		int isw = is(W,start[c]);
		if (isw || (is(B,start[c]) && is(W,start[c+1])))
			--cioccc;
		if (inword)    inword = !isw;
		else if (!isw) inword = ++cword;
	}
}

void compat_kw(int tlen, int inword) {
	int c, w=0;
	for (w=c=2; c<tlen; ++c) {	// skip comment lead-in chars
		int isw = !is(I|D,start[c]);
		if (inword) {
			inword = !isw;
			if (!inword && iskw(start+w, c-w))
				cioccc -= c-w-1;
		} else if (!isw)
			inword = w = c;
	}
}

void compat(int t, int tlen) {
	static int tprev;
	int c;
	if (tprev == CMT) {
		/*
		 * do not bcount spaces following comment blocks,
		 * including the newline!  nor lcount the newline following
		 * a comment block if last on line
		 */
		unsigned char *ow = white;
		while (ow<start && *ow++ != '\n')
			;
		if (!fkeep && ow[-1] == '\n')
			--cline;
		cbyte -= ow-white;
		tprev = 0;
	}
	// new-style comments eat preceding newline,
	// leaving no whitespace if token starts line
	// add it back here for compat word boundary detection
	if (white == start && white[-1] == '\n')
		--white;
	if (t == CMT) {
		// do not count newlines within comments
		// and checks for keywords in new-style comments
		if (fkeep) {
			int inword = white==start && start>buf;
			cioccc += tlen;
			if (start[1] == '/')
				compat_kw(tlen, inword);
			compat_words(tlen, inword);
		} else for (c=0; c<tlen; ++c)
			if (start[c] == '\n') --cline;

		// nor spaces before comments IFF the comment
		// is first non-white on line
		unsigned char *os = start;
		while (os > white) {
			if (*--os == '\n')
				break;
		}

		// look for last ends with newline,
		// only remove ccounts if whitespace
		if (*os == '\n' || os == buf || white[-1] == '\n')
			cbyte -= start - os - (*os == '\n');

		// look for whitespace before the comment, as well as after
		if (!fkeep
		&& (white != start || start == buf)
		&& !is(W, start[tlen]) && start[1] == '*')
			++cword;

		if (start[1] == '*')
			tprev = CMT;
	} else if (t == '"' || t == '\'') {
		// iocccsize -i does not count whitespace WITHIN a string!!!
		// and {;} can be "free" within a string if follows the "rules"
		compat_words(tlen, 1);
	}
	// count words, not tokens (a preceding old-style CMT is whitespace)
	if (t && t != CMT && white == start && start > buf)
		--cword;
	if (!fsuppress && (t != CMT || fkeep))
		printf("%.*s", (int)(start-white)+tlen, white);
}

void count() {
	int t, tlen;
	kwinit();
	while ((t=lex())) {
		tlen = cur-start;
		if (t != CMT) {
			cbyte += tlen;
			++cword;
			if (freserved && t == KEYW)
				cioccc += 1;
			else if (t < 256 && is(B, t) && (is(W, *cur) || !*cur))
				;
			else
				cioccc += tlen;
		}
		if (fcompat)
			compat(t, tlen);
		else if (!fsuppress && (t != CMT || fkeep)) {
			if (!ftoken) printf("%3d ", t);
			printf("%.*s", tlen, start);
			if (cur[-1] != '\n') printf("\n");
		}
	}

	if (fcompat) {
		// process comments just before end of file
		compat(0, end-white);
	}

	if (fkeep) cbyte = end - buf;
	if (!freserved) tkey = 0;
	tsaved = cbyte - cioccc - tkey;
	cbyte += cextra;
}

int main(int ac, char **av) {
	static char usage[] = "[flags] [file]\n\
	-t  tokenize input\n\
	-c  compatible count and display token ids\n\
	-k  keep comments\n\
	-s  suppress source\n\
	-r  reserved words count as 1\n\
	-i  ioccc size (implies -rs)\n\
	-h  this help";

	fcompat = 1;	// default: compact with iocccsize: invert sense of -c

	int c;
	while ((c = getopt(ac, av, "?tcksrih")) != EOF) {
	    switch (c) {
	    case 't': ftoken    = 1; // fallthru - -t implies -c
	    case 'c': fcompat   = 0; break;
	    case 'k': fkeep     = 1; break;
	    case 's': fsuppress = 1; break;
	    case 'r': freserved = 1; break;
	    case 'i': fioccc = freserved = fsuppress = fcompat = 1; break;
	    case 'h': // fallthru, alias for -?
	    default:
	    	fprintf(stderr, "%s %s\n", av[0], usage);
		return 1;
	    }
	}

	ac -= optind;
	av += optind;
	if (ac > 0 && !freopen(*av, "r", stdin)) {
		perror(*av);
		exit(1);
	}

	readsource();
	count();

	if (!ftoken) {
		if (fioccc) printf("%d\n", cioccc);
		else        fprintf(stderr, "%d %d %d %d %d %d %d\n",
				cline,cword,cbyte,cioccc,tsaved,ckey,tkey);
	}
	return 0;
}
